src/Lecture2.hs

```diff
@@ -292,4 +441,72 @@ Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding expr =
+  let
+    decomposeExpr = exprDecompose expr
+  in uncurry exprCompose decomposeExpr
+
+exprDecompose :: Expr -> ([Expr], [Expr])
```

Comment:
The same advice about boolean blindless applies here. Instead of checking `isVar` and similar to you can use pattern matching.

Additionally, you can decompose expression into `([String], Int)` instead of `([Expr], [Expr])` as a more precise type.